channel/contractbackend.go

```diff
@@ -62,25 +62,116 @@ type Transactor interface {
 // This is needed to send on-chain transaction to interact with the smart contracts.
 type ContractBackend struct {
 	ContractInterface
-	tr                Transactor
+	tr              Transactor
+	noncer          noncer
+	txFinalityDepth uint64
+	chainID         ChainID
+}
+
+type noncer interface {
+	setNonce(chainID ChainID, addr common.Address, nonce uint64) (uint64, error)
+}
+
+// LocalNoncer implements noncer interface.
+type LocalNoncer struct {
+	expectedNextNonce map[common.Address]uint64
 	nonceMtx          *sync.Mutex
+}
+
+// GlobalNoncer implements a global nonce counter that is shared across different contract backends.
+type GlobalNoncer struct {
 	expectedNextNonce map[common.Address]uint64
-	txFinalityDepth   uint64
-	chainID           ChainID
+	nonceMtx          map[common.Address]*sync.Mutex
+}
+
+func (d GlobalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
+	if d.nonceMtx[sender] == nil {
+		d.nonceMtx[sender] = &sync.Mutex{}
+	}
+	d.nonceMtx[sender].Lock()
+	defer d.nonceMtx[sender].Unlock()
+
+	expectedNextNonce, found := d.expectedNextNonce[sender]
+
+	if !found {
+		d.expectedNextNonce[sender] = 0
+	}
+
+	// Compare nonces and use larger.
+	if nonce < expectedNextNonce {
+		nonce = expectedNextNonce
+	}
+
+	// Update local expectation.
+	d.expectedNextNonce[sender] = nonce + 1
+
+	return nonce, nil
+}
+
+func (d LocalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
```

Comment:
```suggestion
func (d *LocalNoncer) setNonce(chainID ChainID, sender common.Address, nonce uint64) (uint64, error) {
```

Applying my comment from my small essay regarding the `GlobalNoncer` would require this to be a pointer receiver as well.